src/aspire/orientation/commonline_sync.py

```diff
@@ -1,31 +1,417 @@
+            cl_idx31 = clmatrix[k_list, i]
+            cl_idx23 = clmatrix[j, k_list]
+            cl_idx32 = clmatrix[k_list, j]
+
+            # C1, C2, and C3 are unit circles of image i, j, and k
+            # theta1 is the angle on C1 created by its intersection with C3 and C2.
+            # theta2 is the angle on C2 created by its intersection with C1 and C3.
+            # theta3 is the angle on C3 created by its intersection with C2 and C1.
+            theta1 = (cl_idx13 - cl_idx12) * 2 * np.pi / n_theta
+            theta2 = (cl_idx21 - cl_idx23) * 2 * np.pi / n_theta
+            theta3 = (cl_idx32 - cl_idx31) * 2 * np.pi / n_theta
+
+            c1 = np.cos(theta1)
+            c2 = np.cos(theta2)
+            c3 = np.cos(theta3)
+
+            # Each common-line corresponds to a point on the unit sphere. Denote the
+            # coordinates of these points by (Pix, Piy Piz), and put them in the matrix
+            #   M=[ P1x  P2x  P3x ;
+            #       P1y  P2y  P3y ;
+            #       P1z  P2z  P3z ].
+            #
+            # Then the matrix
+            #   C=[  1  c1  c2 ;
+            #       c1   1  c3 ;
+            #       c2  c3   1 ],
+            # where c1, c2, c3 are given above, is given by C = M.T @ M.
+            # For the points P1, P2, and P3 to form a triangle on the unit sphere, a
+            # necessary and sufficient condition is for C to be positive definite. This
+            # is equivalent to
+            #       1+2*c1*c2*c3-(c1^2+c2^2+c3^2) > 0.
+            # However, this may result in a triangle that is too flat, that is, the
+            # angle between the projections is very close to zero. We therefore use the
+            # condition below
+            #       1+2*c1*c2*c3-(c1^2+c2^2+c3^2) > 1.0e-5.
+            # This ensures that the smallest singular value (which is actually
+            # controlled by the determinant of C) is big enough, so the matrix is far
+            # from singular. This condition is equivalent to computing the singular
+            # values of C, followed by checking that the smallest one is big enough.
+            cond = 1 + 2 * c1 * c2 * c3 - (
+                    np.square(c1) + np.square(c2) + np.square(c3))
+
+            good_idx = np.where(cond > 1e-5)[0]
+
+            # Calculated cos values of angle between i and j images
+            cos_phi2 = (c3[good_idx] - c1[good_idx] *
+                        c2[good_idx]) / (np.sin(theta1[good_idx])
+                                         * np.sin(theta2[good_idx]))
+            if np.any(np.abs(cos_phi2) - 1 > 1e-12):
+                logger.warning(f'Globally Consistent Angular Reconstruction (GCAR) exists'
+                               f' numerical problem: abs(cos_phi2) > 1, with the'
+                               f' difference of {np.abs(cos_phi2)-1}.')
+            cos_phi2 = np.clip(cos_phi2, -1, 1)
+            phis[:len(good_idx)] = cos_phi2
```

Comment:
There's no need to declare these `phis` and `inds` arrays ahead of time. We can just set them here:
```
phis = cos_phi2
inds = k_list[good_idx]
```

Reply:
Declarations are removed in commit 7b420f9.